tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -365,7 +371,7 @@ bool accessSubscriptsAreUnrolledLoops(
     auto subdomain = activeDomainPointsBelow(root, leaf);
 
     auto unrolledDims = isl::union_pw_aff_list(leaf->ctx_, 1);
-    for (auto node : ancestors) {
+    for (const detail::ScheduleTree* node : ancestors) {
       auto band = node->elemAs<detail::ScheduleTreeElemBand>();
```

Comment:
Writing this as `auto band = node->template elemAs<detail::ScheduleTreeElemBand>();` allows me to keep the `auto`, but it is again not obvious why I need to introduce `template` here.

Reply:
Syntactic disambiguation.  A particularly nasty interplay between templates and type inference.  The syntax of a template is checked when parsing its definition.  But type inference for `auto` is performed at template instantiation time.  When parsing `elemAs<detail::ScheduleTreeElemBand`, the compiler wouldn't know if  `<` is a comparison operator or the start of a template argument list.  Unless `elemAs` is known to be a template (but it is not known because it's type has not been inferred yet when `node` is declared as `auto`), `<` is treated as a comparison operator.  In this case, the RHS must be an expression. `detail::ScheduleTreeElemBand>()` is not a valid expression so the compiler complains.  Adding the `template` prefix indicates that `<` must be always treated as the start of the template argument list, so it fixes the problem.  So does manually specifying the type of `node` because the compiler now knows that `elemAs` is a function template.